File: stores/itunes.js
const fetch = require('node-fetch');

const STORE_NAME = 'iTunes / Apple Music';
const STORE_URL = 'https://music.apple.com';

async function search(query) {
  try {
    const url = `https://itunes.apple.com/search?term=${encodeURIComponent(query)}&media=music&entity=song&limit=25`;
    const res = await fetch(url, { timeout: 10000 });
    if (!res.ok) return [];

    const data = await res.json();
    return (data.results || []).map(track => ({
      title: track.trackName,
      artist: track.artistName,
      label: track.collectionName || '',
      genre: track.primaryGenreName || '',
      bpm: null,
      key: null,
      duration: track.trackTimeMillis ? formatDuration(track.trackTimeMillis) : '',
      price: track.trackPrice ? `${track.trackPrice} ${track.currency}` : 'N/A',
      priceValue: track.trackPrice || null,
      currency: track.currency || 'USD',
      artwork: track.artworkUrl100 ? track.artworkUrl100.replace('100x100', '200x200') : null,
      url: track.trackViewUrl,
      store: STORE_NAME,
      storeIcon: 'apple',
      releaseDate: track.releaseDate ? track.releaseDate.split('T')[0] : '',
    }));
  } catch (err) {
    console.error(`iTunes search error: ${err.message}`);
    return [];
  }
}

function formatDuration(ms) {
  const mins = Math.floor(ms / 60000);
  const secs = Math.floor((ms % 60000) / 1000);
  return `${mins}:${secs.toString().padStart(2, '0')}`;
}

module.exports = { search, STORE_NAME };

File: stores/itunes.py
from __future__ import annotations

from urllib.parse import quote

import requests

from models import TrackResult

STORE_NAME = "iTunes / Apple Music"


def _format_duration(ms: int) -> str:
    mins = ms // 60000
    secs = (ms % 60000) // 1000
    return f"{mins}:{secs:02d}"


def search(query: str) -> list[TrackResult]:
    try:
        url = f"https://itunes.apple.com/search?term={quote(query)}&media=music&entity=song&limit=25"
        resp = requests.get(url, timeout=10)
        if not resp.ok:
            return []

        data = resp.json()
        results: list[TrackResult] = []

        for track in data.get("results", []):
            duration = ""
            if track.get("trackTimeMillis"):
                duration = _format_duration(track["trackTimeMillis"])

            price = "N/A"
            price_value = None
            if track.get("trackPrice"):
                price = f"{track['trackPrice']} {track.get('currency', 'USD')}"
                price_value = track["trackPrice"]

            artwork = None
            if track.get("artworkUrl100"):
                artwork = track["artworkUrl100"].replace("100x100", "200x200")

            release_date = ""
            if track.get("releaseDate"):
                release_date = track["releaseDate"].split("T")[0]

            results.append(TrackResult(
                title=track.get("trackName", ""),
                artist=track.get("artistName", ""),
                label=track.get("collectionName", ""),
                genre=track.get("primaryGenreName", ""),
                bpm=None,
                key=None,
                duration=duration,
                price=price,
                price_value=price_value,
                currency=track.get("currency", "USD"),
                artwork=artwork,
                url=track.get("trackViewUrl", ""),
                store=STORE_NAME,
                store_icon="apple",
                release_date=release_date,
            ))

        return results
    except Exception as e:
        print(f"iTunes search error: {e}")
        return []
